### PR TITLE
feat: Folder Explorer as an independent component

### PR DESCRIPTION
**Pull request by ironAiken2**
### TL;DR

This is a task to decouple folder-explorer from storage-list so that folder-explorer can be called and used from anywhere in React Root.

### What changed?

The folder explorer that was dependent on `backend-ai-storage-list` has been separated and used inside `LegacyFolderExplorer`, which can be called and used from any part of react root.

The API response to get files of a vfolder exists files(for legacy compatibility) and items. In explorer, I changed to using items.

Currently, you can use the folder ID passed as a query parameter by passing it as a prop. 

### How to test?

1. navigate to the Data & Storage tab. 
2. create a folder or open an existing folder. 
3. Check the folder explorer modal.

The `LegacyFolderExplorer` provides the same functionality as before. I'd appreciate it if you could check if there are any changes I missed compared to the existing modal.

### Why make this change?

To improve the usability and functionality of the modal and folder explorer components in the application.

---

**Checklist:** (if applicable)

- [ ] Mention to the original issue
- [ ] Documentation
- [ ] Minium required manager version
- [ ] Specific setting for review (eg., KB link, endpoint or how to setup)
- [ ] Minimum requirements to check during review
- [ ] Test case(s) to demonstrate the difference of before/after

### CODE REVIEW

**Comment by ironAiken2** [comment]
* **#2529** <a href="https://app.graphite.dev/github/pr/lablup/backend.ai-webui/2529?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @ironAiken2 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by rapsealk** [comment]
Please check for the title! 😉

**Review by yomybaby — CHANGES_REQUESTED** [review]
Please check below:
- In the existing version, the delete button at the top appears when the user selects files.
    
- When the explorer is closed, the selected tab in the query string is deleted.


<div class='graphite__hidden'>
          <div>🎥 Video uploaded on Graphite:</div>
            <a href="https://app.graphite.dev/media/video/XqC2uNFuj0wg8I60sMUh/b89e4e27-df48-44b1-8a15-428a8d6cf1e2.mp4">
              
            </a>
          </div>
<video src="https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/XqC2uNFuj0wg8I60sMUh/b89e4e27-df48-44b1-8a15-428a8d6cf1e2.mp4">iShot_2024-07-16_12.44.42.mp4</video>

File: src/components/backend-ai-data-view.ts [hunk 1]
```diff
@@ -827,6 +827,7 @@ export default class BackendAIData extends BackendAIPage {
   }
 
   openFolderExplorer = (e) => {
+    console.log('Event: ', e);
```
Comment (yomybaby) [inline src/components/backend-ai-data-view.ts]: Please delete this

File: src/components/backend-ai-storage-list.ts
Comment (yomybaby) [inline src/components/backend-ai-storage-list.ts:1]: Please delete the code related to the folder explorer that is not currently being used.

Comment (yomybaby) [reply src/components/backend-ai-storage-list.ts:1]: And please check the changes in these files. It looks like there is a mistake.

**Review by yomybaby — CHANGES_REQUESTED** [review]

File: src/components/backend-ai-storage-list.ts
Comment (ironAiken2) [reply src/components/backend-ai-storage-list.ts:1]: I don't know why this file got a modification.... I've restored it to its original state.

**Comment by ironAiken2** [comment]
All requests have been applied. @yomybaby

**Review by agatha197 — CHANGES_REQUESTED** [review]
1. The previous folder name remained after creating/renaming a new sub folder(create inside the folder explorer).
2. I cannot delete subfolders

3. I cannot download any files.
4. When uploading a folder, it overwrites regardless of case sensitivity. For example, if you create a folder named 'a' in the file explorer and upload a folder named 'A', the files inside the 'A' folder will be placed inside the 'a' folder.
5. When I renamed a file, I got this error from storage proxy.

6. The download icon is sometimes inactive even though the permission is enabled.

**Comment by ironAiken2** [comment]
2. I cannot delete subfolders 
-> Fixed.  The issue was caused by not being able to read the filename.

File: react/src/components/FolderExplorerOpener.tsx [hunk 2]
```diff
@@ -38,22 +41,33 @@ const FolderExplorerOpener = () => {
...
-            }),
-          );
+          // document.dispatchEvent(
+          //   new CustomEvent('folderExplorer:open', {
+          //     detail: {
+          //       vFolder,
+          //     },
+          //   }),
+          // );
+          setVFolderName(vFolder?.name || '');
```
Comment (yomybaby) [inline react/src/components/FolderExplorerOpener.tsx]:
```suggestion
          setVFolderName(vFolder?.name || '');
```

File: react/src/components/BAIModal.tsx [hunk 3]
```diff
@@ -46,7 +51,7 @@ const BAIModal: React.FC<BAIModalProps> = ({ styles, ...modalProps }) => {
         keyboard={false}
         {...modalProps}
         centered={modalProps.centered ?? true}
-        className="bai-modal"
+        className={`bai-modal ${className ?? ''}`}
```
Comment (yomybaby) [inline react/src/components/BAIModal.tsx:54]: Next time, let's consider using https://github.com/JedWatson/classnames.

File: src/components/backend-ai-folder-explorer.ts [hunk 4]
```diff
@@ -0,0 +1,1552 @@
...
+        } else {
+          return {
+            valid: nativeValidity.valid,
+            customError: !nativeValidity.valid,
+          };
+        }
+      } else {
+        // custom validation for path name using regex
+        const regex =
+          /^([^`~!@#$%^&*()|+=?;:'",<>{}[\]\r\n/]{1,})+(\/[^`~!@#$%^&*()|+=?;:'",<>{}[\]\r\n/]{1,})*([/,\\]{0,1})$/gm;
```
[the remark this answers is not in the document]
Comment (agatha197) [reply src/components/backend-ai-folder-explorer.ts]: suggest: ```^[^`~!@#$%^&*()|+=?;:'",<>{}[\]\r\n/]+(\/[^`~!@#$%^&*()|+=?;:'",<>{}[\]\r\n/]+)*[\/\\]?$```

**Review by agatha197 — CHANGES_REQUESTED** [review]
1. It would be better to show this modal if user try to modify extension.

2. And the bottom buttons of upper confirm modal doesn't have `gap`.
3. The right bottom button string is too long.

**Review by agatha197 — CHANGES_REQUESTED** [review]
Let's add `destroyOnClose` to rename modal.

**Comment by ironAiken2** [comment]
> Let's add destroyOnClose to rename modal.

Since the modal is written as a web component, I modified it to set empty the field when the modal opens.
@agatha197

**Review by agatha197 — APPROVED** [review]
LGTM!

**Review by yomybaby — CHANGES_REQUESTED** [review]

**Review by yomybaby — APPROVED** [review]
🙌 LGTM with #2622. Don't merge without #2622 .

**Pull request merged**